packages/stream/index.js

```diff
@@ -0,0 +1,147 @@
+/* `makeQueue`, `makeStream`, and `makePipe` are utilities for creating async
+ * iterator "streams". A Stream is compatible with AsyncIterator and Generator
+ * but differ in that every method and argument of both is required.
+ * For example, streams always have `return` and `throw` for closing the write
+ * side.
+ * The `Stream` interface is symmetric, but a stream that sends data and
+ * receives undefined is conventionally a `Writer` whereas a stream that
+ * receives data and sends undefined is conventionally a `Reader`.
+ */
+
+// @ts-check
+/// <reference types="ses"/>
+
+/**
+ * @template T
+ * @typedef {{
+ *   resolve(value?: T | Promise<T>): void,
+ *   reject(error: Error): void,
+ *   promise: Promise<T>
+ * }} PromiseKit
+ */
+
+/**
+ * @template T
+ * @returns {PromiseKit<T>}
+ */
+const makePromiseKit = () => {
+  let resolve;
+  let reject;
+  const promise = new Promise((res, rej) => {
+    resolve = res;
+    reject = rej;
+  });
+  assert(resolve !== undefined);
+  assert(reject !== undefined);
+  return harden({ promise, resolve, reject });
+};
+
+/**
+ * @template T
+ * @typedef {{
+ *   put(value: T | Promise<T>): void,
+ *   get(): Promise<T>
+ * }} AsyncQueue
+ */
+
+/**
+ * @template T
+ * @returns {AsyncQueue<T>}
+ */
+export const makeQueue = () => {
+  let { promise: tailPromise, resolve: tailResolve } = makePromiseKit();
+  return {
+    put(value) {
+      const { resolve, promise } = makePromiseKit();
+      tailResolve(Object.freeze({ value, promise }));
+      tailResolve = resolve;
+    },
+    get() {
+      const promise = tailPromise.then(next => next.value);
+      tailPromise = tailPromise.then(next => next.promise);
```

Comment:
Are we trying to avoid depending on eventual-send / HandledPromise? I ask because I'd normally write this as
```suggestion
      const promise = E.get(tailPromise).value;
      tailPromise = E.get(tailPromise).promise;
```